components/zcash_address/src/encoding.rs

```diff
@@ -0,0 +1,298 @@
+use std::{convert::TryInto, error::Error, fmt, str::FromStr};
+
+use bech32::{self, FromBase32, ToBase32, Variant};
+
+use crate::{kind::*, AddressKind, Network, ZcashAddress};
+
+/// An error while attempting to parse a string as a Zcash address.
+#[derive(Debug, PartialEq)]
+pub enum ParseError {
+    /// The string is an invalid encoding.
+    InvalidEncoding,
+    /// The string is not a Zcash address.
+    NotZcash,
+}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ParseError::InvalidEncoding => write!(f, "Invalid encoding"),
+            ParseError::NotZcash => write!(f, "Not a Zcash address"),
+        }
+    }
+}
+
+impl Error for ParseError {}
+
+impl FromStr for ZcashAddress {
+    type Err = ParseError;
+
+    /// Attempts to parse the given string as a Zcash address.
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Remove leading and trailing whitespace, to handle copy-paste errors.
+        let s = s.trim();
+
+        // Most Zcash addresses use Bech32, so try that first.
+        match bech32::decode(s) {
+            Ok((hrp, data, Variant::Bech32m)) => {
+                // If we reached this point, the encoding is supposed to be valid Bech32m.
+                let data =
+                    Vec::<u8>::from_base32(&data).map_err(|_| ParseError::InvalidEncoding)?;
+
+                let net = match hrp.as_str() {
+                    unified::MAINNET => Network::Main,
+                    unified::TESTNET => Network::Test,
+                    unified::REGTEST => Network::Regtest,
+                    // We will not define new Bech32m address encodings.
```

Comment:
I think it's possible that we will. For instance, we may find a way to reduce the overhead currently imposed by the 16 bytes of redundancy, or some other way to mitigate address replacement attacks that doesn't use F4Jumble. Or there could be some cryptographic flaw in that design.

Reply:
Mmk, I'll remove this comment. But since there's no way to predict how we might make that kind of change, I can't do what I was previously doing (returning `ParseError::MaybeZcash`).

Reply:
I think that we should reserve another Bech32m HRP for such purposes (just one should be sufficient).